feat(no-await-in-condition): new rule to report use of await within chai-as-promised conditions

brettz9 opened the pull request:
Builds on #9.

feat(no-await-in-condition): new rule to report use of `await` within chai-as-promised conditions

BREAKING CHANGE:

Adds new rule to `recommended`

brettz9 committed:
refactor: use a few es6 features, applying linting rules; also move let closer to scope

Also adds `.eslintignore` (when introspecting on `node_modules` files in IDE, won't complain)

brettz9 committed:
feat(no-await-in-condition): new rule to report use of `await` within chai-as-promised conditions

BREAKING CHANGE:

Adds new rule to `recommended`

dpraul commented on .eslintignore:
```diff
@@ -0,0 +1 @@
+node_modules
```
I don't think this is necessary - eslint should always ignore `node_modules` by default: https://eslint.org/docs/user-guide/configuring#eslintignore

brettz9 replied:
For some time, it seemed to me it wasn't behaving that way, at least in my IDE, but that seems to no longer be the case (and I should know this, as I even recently filed an issue with eslint about their docs being less than clear on this point!) So sure, I plan to remove it shortly.

dpraul commented on lib/rules/no-await-in-condition.spec.js:
```diff
@@ -0,0 +1,78 @@
+    }),
+    innerAwaitExpressions.map(function (code) {
+      return code + '.notify(done)';
+    }),
```
bit of a nitpick, but you can simplify this greatly:
```suggestion
    innerAwaitExpressions.map((code) => `return ${code}`),
    innerAwaitExpressions.map((code) => `await ${code}`),
    innerAwaitExpressions.map((code) => `${code}.notify(done)`),
```

brettz9 replied:
Sure, I can do that. (I was aware of the style, I am just personally not a fan as I can't as easily jump in and log variables, etc.). But we can enforce whatever style and this is of course indeed simple code not needing internal logging anyways.

dpraul commented on lib/rules/no-await-in-condition.spec.js:
```diff
@@ -0,0 +1,78 @@
+      'await Promise.all([' +
+      innerAwaitExpressions.reduce(function(out, expression) {
+        return out + '\n  ' + expression + ',';
+      }, '') +
```
likewise here

```suggestion
      innerAwaitExpressions.reduce((out, expression) => `${out}\n ${expression},`, '') +
```

dpraul commented on test/wrapCodeInTestFunction.js:
```diff
@@ -0,0 +1,5 @@
+
+module.exports = function wrapCodeInTestFunction(code) {
+  return 'it(\'should test\', async function test() { ' + code + '; })';
+};
```
1. Since we're getting some refactoring in, more template strings!
2. I prefer to declare exports separately.

```suggestion
function wrapCodeInTestFunction(code) {
  return `it('should test', async function test() { ${code}; })`;
};

module.exports = wrapCodeInTestFunction;
```

dpraul commented on lib/rules/no-await-in-condition.js:
```diff
@@ -0,0 +1,65 @@
+    });
+    return true;
+  }
+}
```
Please add this to ensure the function has a consistent return-type

```suggestion
  }
  return false;
}
```

dpraul reviewed (changes requested):
Mostly nit-picks, but thanks so much for the addition! And so sorry for taking so long to review it

brettz9 committed:
Update lib/rules/no-await-in-condition.spec.js

Co-Authored-By: Dylan Praul <dpraul@users.noreply.github.com>

brettz9 committed:
Update lib/rules/no-await-in-condition.spec.js

Co-Authored-By: Dylan Praul <dpraul@users.noreply.github.com>

brettz9 committed:
Update test/wrapCodeInTestFunction.js

Co-Authored-By: Dylan Praul <dpraul@users.noreply.github.com>

brettz9 committed:
Update lib/rules/no-await-in-condition.js

Co-Authored-By: Dylan Praul <dpraul@users.noreply.github.com>

brettz9 committed:
- Remove eslintignore

brettz9 committed:
- Lint: Remove unnecessary semi-colon

brettz9 commented:
I've removed the `.eslintignore`, applied the other suggested changes, and made one fix for the linter (removing an unneeded semicolon). I also rebased on `master`.

dpraul reviewed (approved):
Awesome, thanks! LGTM!
